File: BallClock_Scala/src/main/scala/BallClock.scala
import scala.collection.JavaConverters._
import scala.collection.mutable

/**
  * Created by joesturzenegger on 12/11/15.
  */
class BallClock(ballCount: Int) {

  var halfDayCount = 0

  //main ball array
  var mainQueue = createBallList(ballCount)
  var initQueue = createBallList(ballCount)


  //arrays representing the tracks
  var mins = List[Int]()
  var fiveMins = List[Int]()
  var hours = List[Int]()



  def createBallList(count: Int): List[Int] = {
    var queue = List[Int]()
    for (num <- 1 to count) {
      queue = queue :+ num
    }
    queue
  }
  def runClock{
    do {
      addMinute(mainQueue(0))
      mainQueue = mainQueue.drop(1)
    }while(true != checkIfBackAtStart)
    println(ballCount + " balls cycle after " + halfDayCount/2 + " days.")
  }

  def addMinute(ball: Int): Unit = {
    if (mins.length < 4){
      mins = mins :+ ball
    }
    else {
      emptyMins
      addFiveMinute(ball)
    }
  }

  def addFiveMinute(ball: Int): Unit = {
    if (fiveMins.length < 11){
      fiveMins = fiveMins :+ ball
    }
    else {
      emptyFiveMins
      addHour(ball)
    }
  }

  def addHour(ball: Int): Unit = {
    if (hours.length < 11){
      hours = hours :+ ball
    }
    else {
      emptyHours
      mainQueue = mainQueue :+ ball
      halfDayCount += 1;
    }
  }

  def emptyMins: Unit = {
    while(mins.length > 0) {
      mainQueue = mainQueue :+ mins.last
      mins = mins.dropRight(1)
    }
  }

  def emptyFiveMins: Unit = {
    while(fiveMins.length > 0) {
      mainQueue = mainQueue :+ fiveMins.last
      fiveMins = fiveMins.dropRight(1)
    }
  }

  def emptyHours: Unit = {
    while(hours.length > 0) {
      mainQueue = mainQueue :+ hours.last
      hours = hours.dropRight(1)
    }
  }

  def checkIfBackAtStart: Boolean = {
    if(mainQueue.size != initQueue.size) {
      return false
    }
    for(ballNum <- 0 until (ballCount-1)){
      if(mainQueue(ballNum) != initQueue(ballNum)){
        return false
      }
    }
    return true
  }
}

File: BallClock_Swift/BallClock/BallClock.swift
//
//  BallClock.swift
//  BallClock
//
//  Created by Joe Sturzenegger on 5/4/15.
//  Copyright (c) 2015 Joe Sturzenegger. All rights reserved.
//

import Foundation

class BallClock
{
    var ballCount: Int
    var halfDayCount: Int
    
    //main ball array
    var mainQueue: Array<Int>
    
    var initQueue: Array<Int>
    
    //arrays representing the tracks
    var mins: Array<Int>
    var fiveMins: Array<Int>
    var hours: Array<Int>
    
    init(ballCount: Int)
    {
        self.ballCount = ballCount
        halfDayCount = 0
        
        mainQueue = BallClock.createBallArray(ballCount: ballCount)
        initQueue = BallClock.createBallArray(ballCount: ballCount)
        
        mins = [Int]()
        fiveMins = [Int]()
        hours = [Int]()
    }
    
    class func createBallArray(ballCount ballCount:Int)->Array<Int>
    {
        var queue: Array<Int> = [Int]()
        for num in 1...ballCount
        {
            queue.append(num)
        }
        return queue
    }
    
    func runClock()
    {
        repeat
        {
            addMinute(mainQueue.removeAtIndex(0))
        }while !checkIfBackToStart()
        print("\(ballCount) balls cycle after \(Double(halfDayCount)/2) days.")
    }
    func checkIfBackToStart()->Bool
    {
        if mainQueue.count != initQueue.count
        {
            return false
        }
        for ballNum in 0..<ballCount
        {
            if(mainQueue[ballNum] != initQueue[ballNum])
            {
                return false
            }
        }
        return true
    }
    
    func addMinute(ball: Int)
    {
        if mins.count < 4
        {
            mins.append(ball)
        }
        else
        {
            emptyMins()
            addFiveMinute(ball)
        }
    }
    
    func addFiveMinute(ball: Int)
    {
        if fiveMins.count < 11
        {
            fiveMins.append(ball)
        }
        else
        {
            emptyFiveMins()
            addHour(ball)
        }
    }
    
    func addHour(ball: Int)
    {
        if hours.count < 11
        {
            hours.append(ball)
        }
        else
        {
            emptyHours()
            mainQueue.append(ball)
            halfDayCount++;
        }
    }
    func emptyMins()
    {
        
        while mins.count > 0
        {
            mainQueue.append(mins.removeLast())
        }
    }
    
    func emptyFiveMins()
    {
        while fiveMins.count > 0
        {
            mainQueue.append(fiveMins.removeLast())
        }
    }
    
    func emptyHours()
    {
        while hours.count > 0
        {
            mainQueue.append(hours.removeLast())
        }
    }
    
}

